### PR TITLE
Add NewRelic integration

### PR DESCRIPTION
**Pull request by nezorflame**
This is a port of the NewRelic integration which was originally done by @nebolsin in his fork of the tyk [here](https://github.com/nebolsin/tyk/tree/wtf).

We use this changed version of the gateway in production and so far it works as intended, but it'd be really awesome to have this in original tyk without the need for maintaining our fork.

Please review this and feel free to make/request changes.

### CODE REVIEW

**Comment by buger** [comment]
Nice! 

However, i see that it include a lot of refactorings, unrelated to NewRelic. Is it possible to extract only NewRelic part?

**Comment by nezorflame** [comment]
Will work on that.

**Comment by nezorflame** [comment]
@buger still I'd suggest that these refactored parts need to be reviewed as well. Maybe in another pull request?

**Comment by buger** [comment]
I also see that you were trying to solve bugs, caused by router reload logic. I think this is already fixed in our master in a similar manner, so maybe your custom changes will not be needed anymore.

**Comment by nezorflame** [comment]
@buger I see, maybe they won;t be needed anymore then.
Will try only with the NewRelic stuff.

**Comment by nezorflame** [comment]
@buger finished

File: coprocess.go [hunk 1]
```diff
@@ -102,7 +102,10 @@ func (c *CoProcessor) ObjectFromRequest(r *http.Request) *coprocess.Object {
 
 	object.HookType = c.HookType
 
-	object.Metadata = make(map[string]string)
+	object.Metadata = map[string]string{
+		"RequestURI": r.RequestURI,
```
Comment (buger) [inline coprocess.go]: Does it needed for NewRelic integration somehow, or for your internal purpose?

File: reverse_proxy.go [hunk 2]
```diff
@@ -430,6 +431,10 @@ func httpTransport(timeOut int, rw http.ResponseWriter, req *http.Request, p *Re
 		return wsTransport
 	}
 
+	if txn, ok := rw.(newrelic.Transaction); ok {
```
Comment (buger) [inline reverse_proxy.go]: I assume this line should be run only if new relic enabled?

File: middleware.go [hunk 3]
```diff
@@ -53,6 +54,10 @@ func createMiddleware(mw TykMiddleware) func(http.Handler) http.Handler {
 
 	return func(h http.Handler) http.Handler {
 		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			if txn, ok := w.(newrelic.Transaction); ok {
```
Comment (buger) [inline middleware.go]: Same here, maybe we should run it only if new relic is enabled?

File: config/config.go [hunk 4]
```diff
@@ -170,6 +170,12 @@ type SecurityConfig struct {
 	Certificates                     CertificatesConfig `json:"certificates"`
 }
 
+type NewRelicConfig struct {
+	Enabled    bool   `json:"enabled"`
```
Comment (buger) [inline config/config.go]: Do you think it make sense get rid of `Enabled` field, and instead just check if AppName is non empty?

Comment (nezorflame) [reply config/config.go]: OK, can be done.

File: reverse_proxy.go (hunk 2 above)
Comment (nezorflame) [reply reverse_proxy.go]: Yes, will change it.

File: middleware.go (hunk 3 above)
Comment (nezorflame) [reply middleware.go]: Agree.

File: coprocess.go (hunk 1 above)
Comment (nezorflame) [reply coprocess.go]: AFAIK we use this in our events.
If it's possible, it would be nice to leave this as it is.

**Comment by nezorflame** [comment]
Added requested changes.

**Comment by nezorflame** [comment]
@buger anything else to fix/check?

**Comment by nezorflame** [comment]
@buger anything else to fix/check?

**Comment by buger** [comment]
In general looks fine, and it will go through our internal testing. The only thing that concern me is the `RequestURI` part which is not part of NewRelic change. 

I do understand why you added it, and I'm happy to merge it as a second feature (even in terms of this PR), but first let's clarify few things. Do you use it only for JSVM or for Coprocess plugins? I'm asking because JSVM implementation looks good, and coprocess one is a hack, and kind of require adding the new field to proto buffer object. I will create a separate ticket for adding field to coprocess request object anyway, but first, need to know in what kind of middleware you use it right now.

**Comment by nezorflame** [comment]
@buger for coprocess plugins.
We register hooks, and use RequestURI in a function like this:

```go
func (f Flow) ExtractData(ctx coprocess.Context, object *coprocess.Object) (*coprocess.Object, error) {
...
originalURI := object.Metadata["RequestURI"]

	level.Debug(f.logger).Log(
		"msg", "extracted request information",
		"uri", originalURI,
		"env", env,
		"version", version,
		"phone", phone,
		"action", action,
		"key", token,
		"verify", verify,
	)
...
}

**Comment by nezorflame** [comment]
Any news on this?

**Comment by buger** [comment]
We definitely going include this functionality to 2.5, just need to properly implement RequestURI on coprocess side, but this change will be done by us. Will submit PR soon.

**Comment by nezorflame** [comment]
@buger awesome! Could you please @ me when this PR arrives? We'd want to be ready to implement it as soon as it arrives.

**Comment by nezorflame** [comment]
Fixed bug when web transactions weren't sent into NewRelic

**Comment by buger** [comment]
Merged manually https://github.com/TykTechnologies/tyk/commit/2580fd35b24be63e99433d6f8b777b61aa39a455

Note that RequestURI moved from Metadata, to request object field https://github.com/TykTechnologies/tyk/pull/1391

**Comment by nebolsin** [comment]
@nezorflame thanks for pushing this forward 👍

**Pull request closed without merging**